services/rollout-service/pkg/argo/argo_test.go

```diff
@@ -0,0 +1,534 @@
+							},
+							Spec: v1alpha1.ApplicationSpec{
+								Project: "",
+							},
+							Status: v1alpha1.ApplicationStatus{
+								Sync:   v1alpha1.SyncStatus{Revision: "1234"},
+								Health: v1alpha1.HealthStatus{},
+							},
+						},
+					},
+				},
+				{
+					Event: &v1alpha1.ApplicationWatchEvent{
+						Type: "MODIFIED",
+						Application: v1alpha1.Application{
+							ObjectMeta: metav1.ObjectMeta{
+								Name:        "foo",
+								Annotations: map[string]string{},
+							},
+							Spec: v1alpha1.ApplicationSpec{
+								Project: "",
+							},
+							Status: v1alpha1.ApplicationStatus{
+								Sync:   v1alpha1.SyncStatus{Revision: "1234"},
+								Health: v1alpha1.HealthStatus{},
+							},
+						},
+					},
+				},
+				{
+					RecvErr:       status.Error(codes.Canceled, "context cancelled"),
+					CancelContext: true,
+				},
+			},
+			Overview: &api.GetOverviewResponse{
+				Applications: map[string]*api.Application{
+					"foo": {
+						Releases: []*api.Release{
+							{
+								Version:        1,
+								SourceCommitId: "00001",
+							},
+						},
+						Team: "footeam",
+					},
+				},
+				EnvironmentGroups: []*api.EnvironmentGroup{
+					{
+
+						EnvironmentGroupName: "staging-group",
+						Environments: []*api.Environment{
+							{
+								Name: "staging",
+								Applications: map[string]*api.Environment_Application{
+									"foo": {
+										Name:    "foo",
+										Version: 1,
+										DeploymentMetaData: &api.Environment_Application_DeploymentMetaData{
+											DeployTime: "123456789",
+										},
+									},
+								},
+								Priority: api.Priority_UPSTREAM,
+								Config: &api.EnvironmentConfig{
+									Argocd: &api.EnvironmentConfig_ArgoCD{
+										Destination: &api.EnvironmentConfig_ArgoCD_Destination{
+											Name:   "staging",
+											Server: "test-server",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				GitRevision: "1234",
+			},
+			ExpectedReady:    true,
+			ExpectedConsumed: 2,
+		},
+	}
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			as := &mockApplicationServiceClient{
+				Steps:     tc.Steps,
+				cancel:    cancel,
+				t:         t,
+				lastEvent: make(chan *ArgoEvent, 10),
+			}
+			hlth := &setup.HealthServer{}
+			argoProcessor := &mockArgoProcessor{
+				lastOverview:      tc.Overview,
+				ApplicationClient: as,
+				trigger:           make(chan *api.GetOverviewResponse, 10),
+				argoApps:          make(chan *v1alpha1.ApplicationWatchEvent, 10),
+			}
+			hlth.BackOffFactory = func() backoff.BackOff { return backoff.NewConstantBackOff(0) }
+			go argoProcessor.Consume(t, ctx)
+			go argoProcessor.ConsumeArgo(ctx, hlth.Reporter("consume-argo"))
+			argoProcessor.Push(ctx, tc.Overview)
+			//We add a delay so that all the events are reported by the application client
```

Comment:
I force a delay here to make sure the events created from the apps are matched before the consumed events test function is called